Micromagnetics

MSallermann opened the pull request:
Implements the micromagnetic Hamiltonian.

- Closes #362

Todo:
- [ ] implement gradients
- [ ] fully implement CUDA backend
- [ ] implement unit tests
- [ ] implement NIST standard problems, see #296

GPMueller committed:
Core: added Micromagnetic Hamiltonian.
The class does not have any real functionality yet.
A corresponding input parser has been added, but the kind and units of input parameters still need to be decided.

GPMueller committed:
Core: fixed physical constants.
Added Constants_Micromagnetic namespace.
Also added Joule to other constants namespaces.

GPMueller committed:
Core: added stub for chaning the Hamiltonian type.
This will later be used to change it at runtime, especially enabling one to switch between atomistic and micromagnetic calculations.
Also potentially fixed a crashing bug in the copy constructors of Spin_System: the geometry shared pointer of the Hamiltonian was not set correctly on copy.

GPMueller committed:
Core: updated API and tests for constants.

GPMueller committed:
UI-QT: added widget for micromagnetic Hamiltonian.
Also added the ability to switch between Hamiltonians from the GUI.

GPMueller committed:
Bugfixes for switching between Hamiltonians.
- the API now allows switching to Hamiltonian_Heisenberg
- the GUI was not updating the corresponding tab correctly, potentially leading to crashes
- the CUDA backend of Hamiltonian_Micromagnetic can now be compiled

GPMueller committed:
Core: micromagnetic parameters in configparser.

GPMueller committed:
Updated docs with micromagnetic model.

DTolm committed:
Core: added micromagnetic energy gradients.
Anisotropy, exchange and DMI.
The implementation of the energy calculation is tentative.
Also added some logging of  micromagnetic input parser.

DTolm committed:
Hamiltonian micromagnetic
-fixed anisotropy, exchange DMI - passing skyrmion convergence test now

DTolm committed:
Cuda implementation of micromagnetics (no ddi)

GPMueller committed:
Core: Started to implement input parsing for micromagnetic hamiltonian.

GPMueller committed:
Some improvements on usability of micromagnetics.
- moved DDI_Method enum class to Hamiltonian.hpp
- added settings for dipolar interactions to micromagnetic Hamiltonian
- added some micromagnetic parameters to Hamiltonian API and made some minor improvements
- made external field consistent for micromagnetic Hamiltonian

GPMueller committed:
UI-Qt: connected a few micromagnetics settings.

Commit:
Core: Cuda code compiles now. Hamiltonian_Micromagnetic.cu still needs to be fixed.

Commit:
Core: Made Hamiltonian_Micromagnetic.cu consistent with Hamiltonian_Micromagnetic.cpp

DTolm committed:
Core: CUDA DDI implementation for micromagnetics

GPMueller committed:
Merge branch 'develop' into fixed-micromagnetic

GPMueller committed:
Merge branch 'develop' into fixed-micromagnetic

GPMueller committed:
Core: compilation fixes in `Hamiltonian_Micromagnetic`.

GPMueller committed:
Core: applied clang-format to Hamiltonian_Micromagnetic files.

Commit:
Hamiltonian_Micromagnetic: Implemented magnetostatics

Commit:
Improved how the demag tensor implementation handles edge cases.

Commit:
Demagnetization_Tensor: Decreased sensitivity for switching to asymptote

Commit:
Hamiltonian_Micromagnetic: Implemented Gradient_DDI_Direct

Commit:
Hamiltonian_Micromagnetic: Moved definition of Ms out of the Hamiltonian (to match mu_s  in atomistic case).
Also added cell_volume to Geometry and changed definition of cell_size (from bounds_max - bounds_min to actual cell_size). Now these fields might also have some use in the atomistic case.

Commit:
Micromagnetics: Current Snapshot

Commit:
Incorporated upstream changes in core/data

Commit:
Incorporated upstream changes in core/engine

Commit:
Incorporated upstream changes in core/io

Commit:
forgotten update of FFT.cu

Commit:
Incorporated upstream changes in core/Spirit

Commit:
Incorporated upstream changes in system.py

Commit:
Incorporated upstream changes in core/utility

Commit:
Merge branch 'develop' into fixed_micromagnetic

Commit:
Fixed some errors stemming from previous merge

Commit:
Hamiltonian_Micromagnetic: Compilation fix

GPMueller committed:
Merge branch 'develop' into fixed-micromagnetic

GPMueller committed:
Core: fix Rydberg/eV ratio and test.

Commit:
Hamiltonian_Micromagnetic: Updated E_DMI term. But there seems to be a factor 2 difference to OOMMF still. Also corrected errors in set_dmi API function.

Commit:
Fix compilation with OpenMP. Wrong number of loops in one of t he collapse clauses.

Commit:
Hamiltonian_Micromagnetic: Fixed errors when compiling with scalar=float

Commit:
Unit-Tests: The test cases "Finite Differences" and "Dipole-Dipole Interaction" no also cover the micromagnetic Hamiltonian.

Commit:
Hamiltonian_Micromagnetic: Corrected Gradient_Exchange

Commit:
Hamiltonian_Micromagnetic: Additions for DDI
- Implemented E_DDI_Direct
- corrected periodic boundary conditions in Gradient_DDI_Direct
- fixed set_ddi API function

Commit:
Hamiltonian_Micromagnetic: Gradients.
- Improved implementation of Gradient_Exchange
- Implemented Gradient_DMI
- Added both to unit test

GPMueller commented on core/include/engine/Demagnetization_Tensor.hpp:
```diff
@@ -0,0 +1,276 @@
+                        if(std::abs(tmp) > cur_max)
+                            cur_max = std::abs(tmp);
+                        if(std::abs(tmp) > cur_max)
+                            cur_max = std::abs(res);
```
this looks like a copy/paste duplication

GPMueller replied:
Fixed with 7fed5e2c2a4e35f404cd7fe9cdd0b7211898c96b

GPMueller commented on core/docs/Input.md:
```diff
@@ -306,6 +307,80 @@ Pairwise interactions are specified in meV per unique pair \<ij\>,
+
+Note that Spirit only supports rectilinear geometries when using this Hamiltonian.
+
+**Units:**
+
+In the micromagnetic model, Spirit uses SI units:
+- `m` (meter) for distances
+- `J` (Joule) for energies
+- `s` (second) for time
+
+Therefore,
+- `A [J/m]` for exchange stiffness
+- `D [J/m^2]` for DMI
+- `K [J/m^3]` for anisotropy
+- `Ms [A/m]` for saturation magnetisation
```
Maybe we should add a section like this to the Heisenberg Hamiltonian documentation. IMO it it is quite useful to have this overview in the beginning

GPMueller committed:
Core: fix CUDA build by commenting out code

GPMueller commented on core/include/engine/Demagnetization_Tensor.hpp:
```diff
@@ -0,0 +1,276 @@
+
+namespace Engine
+{
+namespace Demagnetization_Tensor
```
Since the file has been newly added, clang-format should be run on it

GPMueller replied:
This has been fixed

GPMueller committed:
Update formatting for changes on this branch

GPMueller committed:
Merge branch 'develop' into fixed-micromagnetic

GPMueller committed:
Merge branch 'develop' into fixed-micromagnetic

GPMueller committed:
Core: minor micromagnetism fixes

 - fix a small copy/paste bug and formatting in
   Demagnetization_Tensor.hpp
 - configparser formatting
 - specify units in comments of `mu_s` API
